web/src/unread_ui.ts

```diff
@@ -52,9 +52,7 @@ export function update_unread_banner(): void {
 }
 
 export function hide_unread_banner(): void {
-    // Use visibility instead of hide() to prevent messages on the screen from
-    // shifting vertically.
-    $("#mark_read_on_scroll_state_banner").toggleClass("invisible", true);
+    $("#mark_read_on_scroll_state_banner").toggleClass("hide", true);
```

Comment:
This comment suggests the fact that this pushes down `bottom_whitespace` in some way; how does that work?

Reply:
so `invisible` just makes the element not visible, but it would still be present in the DOM. However, in case of `hide`, it would be removed from DOM as well.

In this case, when you use `hide` and scroll to the bottom of list, and mark message as unread, the `mark_as_read_banner` would cause the scroll position to slightly shift since a new element is coming up into document. However, in case of `invisible`, nothing would happen since element is already in document -- just that it is made visible.

Reply:
I don't think that's the right terminology -- hide/visible have different behavior for whether it takes up space in the page layout, but the elements are present in the DOM in either case.

Reply:
seems like the `hide` is replaced with `invisible` first in 1b24bfcc3130efd22f8a38fca84a6f5364b55fd5, and the commit message specifies --
>mark as read: Fix banner not closing when x is clicked.
Previously, the X in the mark-as-read banner did nothing.

I don't seem to reproduce this with hide().

Reply:
According to [stackoverflow](https://stackoverflow.com/questions/133051/what-is-the-difference-between-visibilityhidden-and-displaynone):
> display:none means that the tag in question will not appear on the page at all (although you can still interact with it through the dom). There will be no space allocated for it between the other tags.
visibility:hidden means that unlike display:none, the tag is not visible, but space is allocated for it on the page. The tag is rendered, it just isn't seen on the page.

So the comment in the code - 
```
// Use visibility instead of hide() to prevent messages on the screen from
// shifting vertically.
````
maybe refers to shifting of message list due to allocation/deallocation of the space to the `mark as read` banner (in case of `visibility:hidden` since tag is rendered even when hidden, maybe we won't observer the shitft).

However, this issue was not part of #21595 where this was first introduced.

Reply:
Since hiding (`display: none`) the banner instead of making it invisible will reduce scrollable space, it will cause the message list to shift dowards if the scrollbar is at bottom.

To avoid this we can add a `mark_read_on_scroll_state_banner_holder` div below typing notification div which just has the same dimesnsions as the unread banner before it is hidden. Just make sure to keep it hidden by default and hide it when user switches narrows or another unread banner is displayed.

Reply:
Thanks for making that change. Works fine but we can possible make this look better by extracting the call:
```js
    $("#mark_read_on_scroll_state_banner_place_holder").toggleClass("hide", false);
    $("#mark_read_on_scroll_state_banner_place_holder").toggleClass("invisible", true);
```

and call it exactly what it does - `togge_dummy_banner` where we can pass the boolean set `true/false` for these classes.

Reply:
updated!